lib/engine/src/executable.rs

```diff
@@ -0,0 +1,53 @@
+use enumset::EnumSet;
+use wasmer_types::entity::PrimaryMap;
+
+mod private {
+    pub struct Internal(pub(super) ());
+}
+
+/// A WASM module built by some [`Engine`](crate::Engine).
+///
+/// Types implementing this trait are ready to be saved (to e.g. disk) for later use or loaded with
+/// the `Engine` to in order to produce an [`Artifact`](crate::Artifact).
+pub trait Executable {
+    /// The features with which this `Executable` was built.
+    fn features(&self) -> &wasmer_compiler::Features;
+
+    /// The CPU features this `Executable` requires.
+    fn cpu_features(&self) -> EnumSet<wasmer_compiler::CpuFeature>;
+
+    /// The memory styles associated with this `Executable`.
+    fn memory_styles(&self) -> &PrimaryMap<wasmer_types::MemoryIndex, wasmer_vm::MemoryStyle>;
+
+    /// The table plans associated with this `Executable`.
+    fn table_styles(&self) -> &PrimaryMap<wasmer_types::TableIndex, wasmer_vm::TableStyle>;
+
+    /// Data initializers used during module instantiation.
+    fn data_initializers(&self) -> &[wasmer_types::OwnedDataInitializer];
+
+    /// Serializes an artifact into bytes
+    fn serialize(
+        &self,
+        out: &mut dyn std::io::Write,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
```

Comment:
Can we make this return a `Vec` rather than go through Write? Write is of coures more general, but for our use case we don't need it, I believe, and it adds an extra error condition.

Reply:
The reason I wanted to utilize `Write` here was to allow for the upper level to introduce additional data framing. In particular the serialization format here does not maintain any checksums (we _definitely_ should have at least a crc) nor any other similarly potentially useful information. While `Vec` is also workable, doing this while the output is being streamed seemed like a more logical API to achieve this to me.

Can be done in the future though, sure.

Reply:
Yeah, the problem you can't say "I am going to stream, but I also can't fail". That is, I feel it's relatively important to express "serialization can't fail" in the API. I guess the perfect solution would be: 

```
pub fn serialize(&self, buf: &mut Vec<u8>) {
    let _cant_fail = self.serialize_to_writer(&mut buf);
}
pub fn serialize_to_writer(&self, out: &mut dyn Write) -> io::Result {
}
```

Wait, I've just realize that, as stated, this returns a `dyn Error` rather than `io::Error`. This def feels weird to me, there shouldn't be any sources of errors here besides io from `out`, right?

Reply:
This is a trait for arbitrary `Executable`s. Depending on the format, I can definitely see non-I/O serialization failures cropping up.

Previous incarnation of this code had a concrete error type here, but the variants internally were fairly poorly architected IMO, relying heavily on strings to convey information. `Box<dyn std::error::Error>` at least allows callers to attempt downcasting to interesting cases for themselves.

Reply:
Ah, this makes more sense now. Yeah, if we do have an error path here already, than using `Write` is a pure win. But yeah, it'd be good to add some specific API for our use-case (UniversalExecutable), which statically guarantees absence of errors.